File: frontend/src/app/core/services/consumption.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { environment } from '../../../environments/environment';
import { ApiResponse } from '../models/api-response.model';
import { Customer, CustomerListData } from '../models/customer.model';
import { DashboardData, AccountBalance, DataConsumption, MinuteConsumption } from '../models/consumption.model';

@Injectable({ providedIn: 'root' })
export class ConsumptionService {
  private readonly http = inject(HttpClient);
  private readonly baseUrl = `${environment.apiUrl}/bss`;

  getCustomers(): Observable<CustomerListData> {
    return this.http
      .get<ApiResponse<CustomerListData>>(`${this.baseUrl}/customers`)
      .pipe(map(res => res.data));
  }

  getCustomerDashboard(customerId: number): Observable<DashboardData> {
    return this.http
      .get<ApiResponse<DashboardData>>(`${this.baseUrl}/customers/${customerId}/dashboard`)
      .pipe(map(res => res.data));
  }

  getCustomerBalance(customerId: number): Observable<AccountBalance> {
    return this.http
      .get<ApiResponse<AccountBalance>>(`${this.baseUrl}/customers/${customerId}/balance`)
      .pipe(map(res => res.data));
  }

  getCustomerConsumption(customerId: number): Observable<{ data: DataConsumption; minutes: MinuteConsumption }> {
    return this.http
      .get<ApiResponse<{ data: DataConsumption; minutes: MinuteConsumption }>>(
        `${this.baseUrl}/customers/${customerId}/consumption`
      )
      .pipe(map(res => res.data));
  }
}

File: backend/app/services/consumption_service.py
from ..models.customer import Customer
from ..extensions import db


class ConsumptionService:

    @staticmethod
    def get_all_customers() -> list[dict]:
        customers = Customer.query.all()
        return [c.to_dict() for c in customers]

    @staticmethod
    def get_customer_dashboard(customer_id: int) -> dict | None:
        customer = db.session.get(Customer, customer_id)
        if not customer:
            return None
        return {
            'customer': customer.to_dict(),
            'balance': customer.balance.to_dict() if customer.balance else None,
            'data_consumption': customer.data_consumption.to_dict() if customer.data_consumption else None,
            'minute_consumption': customer.minute_consumption.to_dict() if customer.minute_consumption else None,
        }

    @staticmethod
    def get_customer_balance(customer_id: int) -> dict | None:
        customer = db.session.get(Customer, customer_id)
        if not customer or not customer.balance:
            return None
        return customer.balance.to_dict()

    @staticmethod
    def get_customer_consumption(customer_id: int) -> dict | None:
        customer = db.session.get(Customer, customer_id)
        if not customer:
            return None
        return {
            'data': customer.data_consumption.to_dict() if customer.data_consumption else None,
            'minutes': customer.minute_consumption.to_dict() if customer.minute_consumption else None,
        }
